### ISSUE TITLE
Философия код-стайла, настройка преттиера, еслинта, прекоммит-хуки

### ISSUE THREAD

**Issue by rqrqrqrq** [4xxi/frontend-guidelines#9]

**Comment by rqrqrqrq** [comment]
Пришла мысль о том, что в некоторых случаях не стоит делать изменения в прекоммит хуке с последующим `git add`

Описать 2 практики:
1. С фиксами на прекоммите `eslint --fix` `prettier --write`
2. Без фиксов, только проверка соблюдения. Оставить изменения кода на совести разработчиков, их редакторов и специализированных npm-команд.

**Comment by rqrqrqrq** [comment]
Возможно нам стоит полностью перейти на второй путь, избавляет от некоторых проблем.

1. Обновление преттиера/еслинта/etc на проекте => разработчик пуллит обновления и, предположим, у него возникают конфликты мержа, он коммитит мерж => беруться его локальные пакеты старой версии и форматируют не так, как надо. Очень частая проблема при обновлениях, сильно мешает в большой команде.
2. Становится невозможно коммитить файлы частично (по строкам), тк на каждом файле, попавшим в прекоммит, вызывается `git add {{ filename }}`

**Comment by sg-4xxi** [comment]
в целом, я даже за, что то фиксить без своего ведома не айс, пусть будет просто алерт и реджект коммита
а чтобы не запускались глобальные пакеты, можно прямой путь указать в node_modules/.bin проекта

**Comment by rqrqrqrq** [comment]
@sg-4xxi Да, можно, но в нодмодулях старый пакет. Предположим, мы обновили преттиер, разработчик пульнулся и мержит конфликты (но пока не прогнал `yarn`), у него на прекоммите отработает старая версия из нодмодулей

**Comment by sg-4xxi** [comment]
может yarn install на прекоммите? :)

**Comment by rqrqrqrq** [comment]
@sg-4xxi Мне кажется, реджект коммита заставит разработчика прогнать `yarn`. Можно поресерчить в сторону кастомного сообщения при фейле, и писать там напрямую `братан, ты уверен, что ты обновил зависимости?`

В любом случае, даже простой реджет лучше, чем молча отформатировать весь коммит под старую версию

**Comment by rockon404** [comment]
@rqrqrqrq, @sg-4xxi, а почему просто не добавить вызов **yarn** в **post-merge**. Вроде **husky** это поддерживает.

**Comment by rockon404** [comment]
Тупанул. Тут ведь не мерж. Но думаю проблему с **yarn** все таки можно через **husky** разрулить.

**Comment by rqrqrqrq** [comment]
@rockon404 Я не уверен, что можно разрулить не заставляя форсом делать `yarn` при каждом пулле (смотреть, присутствуют при пулле изменения в пакетах?). 

Проблема с добавлением файла в коммит после изменений все равно актуальна (прямо на уровне lint-staged https://github.com/okonet/lint-staged/issues/62)

Ну и я не вижу минусов для тех, у кого настроены редакторы. Для остальных и для специфических случаев, когда упало не там, где ты писал, можно сделать нпм-команды, которые отформатируют всё.

**Comment by rockon404** [comment]
@rqrqrqrq в теории можно запускать по **post-merge** скрипт который запустит
`git diff --name-only HEAD^ HEAD`
и если там будет присутствовать **package.json**, скрипт запустит **yarn**

**Comment by rockon404** [comment]
обсуждение https://gist.github.com/sindresorhus/7996717
